engine.js

```diff
@@ -1,651 +1,655 @@
-'use strict';
-
-var evaluate = require('./evaluate');
-var describe = require('./describe');
-
-function weigh(scope, randomer, expressions, weights) {
-    var weight = 0;
-    for (var i = 0; i < expressions.length; i++) {
-        weights[i] = evaluate(scope, randomer, expressions[i]);
-        weight += weights[i];
-    }
-    return weight;
-}
-
-function pick(weights, weight, randomer) {
-    var offset = Math.floor(randomer.random() * weight);
-    var passed = 0;
-    for (var i = 0; i < weights.length; i++) {
-        passed += weights[i];
-        if (offset < passed) {
-            return i;
-        }
-    }
-    return null;
-}
-
-function pop(array, index) {
-    array[index] = array[array.length - 1];
-    array.length--;
-}
-
-module.exports = class Engine {
-    debug = typeof process === 'object' && process.env.DEBUG_ENGINE
-
-    constructor(args) {
-        this.story = args.story;
-        this.labels = Object.keys(this.story);
-        this.handler = args.handler;
-        this.options = [];
-        this.keywords = {};
-        this.noOption = null;
-        this.global = new Global(this.handler);
-        this.top = this.global;
-        this.start = args.start || 'start';
-        this.label = this.start;
-        this.instruction = {type: 'goto', next: this.start};
-        this.render = args.render;
-        this.dialog = args.dialog;
-        this.dialog.engine = this;
-        this.randomer = args.randomer || Math;
-        this.waypoint = this.capture();
-        Object.seal(this);
-    }
-
-    reset() {
-        this.labels = Object.keys(this.story);
-        this.options = [];
-        this.keywords = {};
-        this.noOption = null;
-        this.global = new Global(this.handler);
-        this.top = this.global;
-        this.label = this.start;
-        this.instruction = {type: 'goto', next: this.start};
-        this.waypoint = this.capture();
+import evaluate from './evaluate.js';
+import describe from './describe.js';
+
+const weigh = (scope, randomer, expressions, weights) => {
+  let weight = 0;
+  for (let i = 0; i < expressions.length; i++) {
+    weights[i] = evaluate(scope, randomer, expressions[i]);
+    weight += weights[i];
+  }
+  return weight;
+};
+
+const pick = (weights, weight, randomer) => {
+  const offset = Math.floor(randomer.random() * weight);
+  let passed = 0;
+  for (let i = 0; i < weights.length; i++) {
+    passed += weights[i];
+    if (offset < passed) {
+      return i;
+    }
+  }
+  return null;
```

Comment:
not for this pr, but this 2-pass sampling algo could be 1-pass; there is no need for the pre-summed weight term; just use `pow(random(), 1/weights[i]`; I guess the "cost" is you go from 1 random call to N, but now you've got a 1-pass online algo fwiw

Reply:
This is a great point and I love the math of it, but also want to avoid floating point because the browsers are constantly whacking behavior divergence moles.